testing/base/source_gen.cpp

```diff
@@ -0,0 +1,690 @@
+  // 12 characters  [1344]  ██████████▊
+  id_length_counts[11] = 12;
+  // 13 characters  [ 498]  ████
+  id_length_counts[12] = 5;
+  // 14 characters  [ 284]  ██▎
+  id_length_counts[13] = 3;
+  // 15 characters  [ 172]  █▍
+  // 16 characters  [ 278]  ██▎
+  // 17 characters  [ 191]  █▌
+  // 18 characters  [ 207]  █▋
+  for (int i = 14; i < 18; ++i) {
+    id_length_counts[i] = 2;
+  }
+  // 19 - 63 characters are all <100 but non-zero, and we map them to 1 for
+  // coverage despite slightly over weighting the tail.
+  for (int i = 18; i < 64; ++i) {
+    id_length_counts[i] = 1;
+  }
+  return id_length_counts;
+}();
+
+// A helper to sum elements of a range.
+template <typename T>
+static auto Sum(const T& range) -> int {
+  return std::accumulate(range.begin(), range.end(), 0);
+}
+
+// A template function that implements the common logic of `GetIds` and
+// `GetUniqueIds`. Most parameters correspond to the parameters of those
+// functions. Additionally, an `AppendIds` callable is provided to implement the
+// appending operation.
+//
+// The main functionality provided here is collecting the correct number of
+// identifiers from each of the lengths in the range [min_length, max_length]
+// and either in our default representative distribution or a uniform
+// distribution.
+//
+// Note that this template must be defined prior to its use below.
+template <typename AppendIds>
+auto SourceGen::GetIdsImpl(int number, int min_length, int max_length,
+                           bool uniform, AppendIds append_ids)
+    -> llvm::SmallVector<llvm::StringRef> {
+  CARBON_CHECK(min_length <= max_length);
+  CARBON_CHECK(uniform || max_length <= 64)
+      << "Cannot produce a meaningful non-uniform distribution of lengths "
+         "longer than 64 as those are exceedingly rare in our observed data "
+         "sets.";
+
+  llvm::SmallVector<llvm::StringRef> ids;
+  ids.reserve(number);
+
+  // First, compute the total weight of the distribution so we know how many
+  // identifiers we'll get each time we collect from it.
+  int num_lengths = max_length - min_length + 1;
+  auto length_counts =
+      llvm::ArrayRef(IdLengthCounts).slice(min_length - 1, num_lengths);
+  int count_sum = uniform ? num_lengths : Sum(length_counts);
+  CARBON_CHECK(count_sum >= 1);
+
+  int number_rem = number % count_sum;
+
+  // Finally, walk through each length in the distribution.
+  for (int length : llvm::seq_inclusive(min_length, max_length)) {
+    // Scale how many identifiers we want of this length if computing a
+    // non-uniform distribution. For uniform, we always take one.
+    int scale = uniform ? 1 : IdLengthCounts[length - 1];
+
+    // Now we can compute how many identifiers of this length to request.
+    int length_count = (number / count_sum) * scale;
+    if (number_rem > 0) {
+      int rem_adjustment = std::min(scale, number_rem);
+      length_count += rem_adjustment;
+      number_rem -= rem_adjustment;
+    }
```

Comment:
I think that
```suggestion
    int length_count = (scale * number) / count_sum;
    if (number_rem > 0) {
      length_count++;
      number_rem--;
    }
```
would give a better approximation of the distribution (with an error of at most one in each bucket's size, rather than at most `IdLengthCounts[i]`).

With the code as-is, if you ask for 120 identifiers with your heuristic distribution, then `count_sum` is something like 350, `number / count_sum` is 0, and you'll end up with 40 of size 1, 40 of size 2, and 40 of size 3, despite size 4 being the largest in the distribution. Whereas with the modified version, you should get around 14 of size 1, 2, and 3, and around 18 of size 4. (Ballpark numbers, I didn't actually work out the correct `count_sum` figure.)

Reply:
So, I tried to reply to the original form of this comment here:

https://github.com/carbon-language/carbon-lang/pull/4124#discussion_r1708920131

I re-checked, and the issue still seems to be there.

Reply:
Sorry, didn't see your previous comment for Github Reasons.

Hm, yeah, I see the problem. It's not easy to predict the amount of rounding error in advance with my approach. We could do it in two passes, but yeah, probably doesn't matter enough to bother given we primarily expect to generate large examples. This is good enough for an initial implementation.